File: 1256-rank-transform-of-an-array/rank-transform-of-an-array.java
class Solution {
    public int[] arrayRankTransform(int[] arr) {
        int n = arr.length;
        int[] sorted = arr.clone();
        Arrays.sort(sorted);
        Map<Integer, Integer> rank = new HashMap<>();
        int r = 1;
        for (int num : sorted) {
            if (!rank.containsKey(num)) {
                rank.put(num, r++);
            }
        }
        int[] result = new int[n];
        for (int i = 0; i < n; i++) {
            result[i] = rank.get(arr[i]);
        }
        return result;
    }
}

File: 1256-rank-transform-of-an-array/rank-transform-of-an-array.cpp
class Solution {
public:
    vector<int> arrayRankTransform(vector<int>& arr) {
        int n = arr.size();
        vector<int> sorted = arr;
        sort(sorted.begin(), sorted.end());
        unordered_map<int, int> rank;
        int r = 1;
        for (int num : sorted) {
            if (!rank.count(num)) {
                rank[num] = r++;
            }
        }
        vector<int> result(n);
        for (int i = 0; i < n; i++) {
            result[i] = rank[arr[i]];
        }
        return result;
    }
};